File: crates/cleat-java/src/main/java/cleat/Memory.java
package cleat;

import java.nio.charset.StandardCharsets;

/**
 * Static helpers for WASM linear memory I/O and bit-packing.
 * <p>
 * Matches the cleat WASM ABI memory layout and bit-packing conventions
 * defined in {@code crates/cleat-sdk/src/memory.rs} and {@code ABI.md}.
 * <p>
 * <h3>Memory Layout</h3>
 * <pre>
 * Offset         | Size    | Use
 * ---------------+---------+-----------------------------------
 * 0xA00000 (10M) | Var     | Scratch region (input args)
 * 0xA10000       | 65536   | Output buffer (results to host)
 * </pre>
 * <p>
 * <h3>TeaVM WASM Memory Access</h3>
 * Under TeaVM's WASM target, linear memory is accessed through
 * {@link org.teavm.interop.Address}.  The SDK uses the scratch region at
 * {@value #SCRATCH_BASE} for packing input arguments to host calls and the
 * output buffer at {@value #OUTPUT_OFFSET} for reading results.
 */
public final class Memory {

    // ---- Constants ----

    /** Size of the output buffer in bytes (64 KiB). */
    public static final int OUT_BUF_SIZE = 65536;

    /** Scratch region base offset in WASM linear memory (10 MiB = 0xA00000). */
    public static final int SCRATCH_BASE = 10 * 1024 * 1024;

    /** Output buffer offset (10 MiB + 64 KiB = 0xA10000). */
    public static final int OUTPUT_OFFSET = SCRATCH_BASE + OUT_BUF_SIZE;

    /**
     * Suspend sentinel value returned by export functions when the workflow
     * needs to suspend (e.g. for a timer or signal).  Equal to {@code 1L << 62}.
     */
    public static final long SUSPEND_SENTINEL = 1L << 62;

    /**
     * Error code for non-retryable terminal errors.
     * Returned as the errCode in {@link #encodeExportResult(int, int)} when
     * a {@link TerminalError} is thrown by the workflow method.
     */
    public static final int TERMINAL_ERROR_CODE = 2;

    // Sleep status constants
    /** Sleep completed normally (replay path). */
    public static final int SLEEP_STATUS_COMPLETED = 0;
    /** Sleep should cause workflow suspension (fresh execution path). */
    public static final int SLEEP_STATUS_SUSPEND = 1;

    private Memory() {
        // Utility class — no instantiation.
    }

    // ---- Raw byte-level memory access via TeaVM Address ----

    /**
     * Read a single byte from WASM linear memory at the given address.
     * <p>
     * Uses {@link org.teavm.interop.Address#fromInt(int)}.{@code getByte()}
     * which TeaVM compiles to an {@code i32.load8_u} WASM instruction.
     *
     * @param address the linear memory offset
     * @return the byte value at that address
     */
    private static byte readByte(int address) {
        return org.teavm.interop.Address.fromInt(address).getByte();
    }

    /**
     * Write a single byte to WASM linear memory at the given address.
     * <p>
     * Uses {@link org.teavm.interop.Address#fromInt(int)}.{@code setByte()}
     * which TeaVM compiles to an {@code i32.store8} WASM instruction.
     *
     * @param address the linear memory offset
     * @param value   the byte value to write
     */
    private static void writeByte(int address, byte value) {
        org.teavm.interop.Address.fromInt(address).putByte(value);
    }

    // ---- String I/O ----

    /**
     * Read a UTF-8 encoded string from WASM linear memory at the given pointer
     * and length.
     *
     * @param ptr the memory offset (inclusive)
     * @param len the number of bytes to read
     * @return the decoded string, or empty string if {@code len <= 0}
     */
    public static String readString(int ptr, int len) {
        if (len <= 0) {
            return "";
        }
        byte[] bytes = new byte[len];
        for (int i = 0; i < len; i++) {
            bytes[i] = readByte(ptr + i);
        }
        return new String(bytes, StandardCharsets.UTF_8);
    }

    /**
     * Write a UTF-8 encoded string to WASM linear memory starting at the given
     * pointer, truncating to a maximum of {@code maxLen} bytes.
     *
     * @param ptr    the memory offset to start writing at
     * @param maxLen the maximum number of bytes to write
     * @param s      the string to write
     * @return the number of bytes actually written
     */
    public static int writeString(int ptr, int maxLen, String s) {
        if (s == null || maxLen <= 0) {
            return 0;
        }
        byte[] bytes = s.getBytes(StandardCharsets.UTF_8);
        int len = Math.min(bytes.length, maxLen);
        for (int i = 0; i < len; i++) {
            writeByte(ptr + i, bytes[i]);
        }
        return len;
    }

    // ---- Export result encoding / decoding ----

    /**
     * Encode an export result into the ABI format:
     * <pre>
     * low 32 bits  = errCode
     * high 32 bits = actualLen
     * </pre>
     *
     * @param errCode   0 for success, non-zero for error
     * @param actualLen number of bytes written to the output buffer
     * @return the packed i64 result
     */
    public static long encodeExportResult(int errCode, int actualLen) {
        return ((long) actualLen << 32) | (errCode & 0xFFFFFFFFL);
    }

    /**
     * Decode the error code from an export result (low 32 bits).
     */
    public static int decodeExportErrCode(long result) {
        return (int) (result & 0xFFFFFFFFL);
    }

    /**
     * Decode the actual length from an export result (high 32 bits).
     */
    public static int decodeExportActualLen(long result) {
        return (int) (result >>> 32);
    }

    // ---- cleat_call result decoding ----
    // bits 40-63 = responseLen (24 bits)
    // bits  8-39 = callErrorCode (32 bits)
    // bits  0-7  = errCode (8 bits)

    /**
     * Decode the response length from a {@code cleat_call} result
     * (bits 40-63).
     */
    public static int decodeCallResponseLen(long result) {
        return (int) ((result >>> 40) & 0xFFFFFFL);
    }

    /**
     * Decode the call error code from a {@code cleat_call} result
     * (bits 8-39).
     */
    public static int decodeCallErrorCode(long result) {
        return (int) ((result >>> 8) & 0xFFFFFFFFL);
    }

    /**
     * Decode the top-level errCode from a {@code cleat_call} result
     * (bits 0-7).  0 = success, 1 = error.
     */
    public static int decodeCallErrCode(long result) {
        return (int) (result & 0xFFL);
    }

    // ---- Simple result decoding (shared by many host calls) ----
    // bits 32-63 = extra
    // bits  0-7  = errCode

    /**
     * Decode the error code from a simple result (bits 0-7).
     * Used by: version, min_version, defer, child_workflow, await_child,
     * continue_as_new.
     */
    public static int decodeSimpleErrCode(long result) {
        return (int) (result & 0xFFL);
    }

    /**
     * Decode the extra value from a simple result (bits 32-63).
     * Used by: now (ms since epoch), random, version, min_version,
     * defer (deferIDLen), child_workflow (runIDLen), await_child (resultLen).
     */
    public static int decodeSimpleExtra(long result) {
        return (int) (result >>> 32);
    }

    // ---- Sleep result decoding ----
    // bits 56-63 = status (0 = completed, 1 = suspend)
    // bits  0-55 = durationMs

    /**
     * Decode the sleep duration from a {@code cleat_sleep} result
     * (bits 0-55).
     */
    public static long decodeSleepDuration(long result) {
        return result & 0x00FFFFFFFFFFFFFFL;
    }

    /**
     * Decode the sleep status from a {@code cleat_sleep} result
     * (bits 56-63).
     *
     * @return {@link #SLEEP_STATUS_COMPLETED} (0) on replay,
     *         {@link #SLEEP_STATUS_SUSPEND} (1) on fresh execution
     */
    public static int decodeSleepStatus(long result) {
        return (int) (result >>> 56);
    }

    // ---- await_signals result decoding ----
    // bits 48-63 = sigNameLen (16 bits)
    // bits 32-47 = payloadLen (16 bits)
    // bits 16-31 = timedOut (1 byte)
    // bits  0-15 = errCode (16 bits)

    /**
     * Decode the signal name length from {@code cleat_await_signals}
     * (bits 48-63).
     */
    public static int decodeAwaitSigNameLen(long result) {
        return (int) ((result >>> 48) & 0xFFFFL);
    }

    /**
     * Decode the payload length from {@code cleat_await_signals}
     * (bits 32-47).
     */
    public static int decodeAwaitPayloadLen(long result) {
        return (int) ((result >>> 32) & 0xFFFFL);
    }

    /**
     * Decode the timedOut flag from {@code cleat_await_signals}
     * (bits 16-31).
     *
     * @return true if the timeout expired before a signal was received
     */
    public static boolean decodeAwaitTimedOut(long result) {
        return ((result >>> 16) & 0xFFFFL) != 0;
    }

    /**
     * Decode the error code from {@code cleat_await_signals}
     * (bits 0-15).
     */
    public static int decodeAwaitErrCode(long result) {
        return (int) (result & 0xFFFFL);
    }

    // ---- await_promise result decoding ----
    // bits 32-63 = resultLen (reuses decodeSimpleExtra)
    // bits 16-23 = timedOut (8 bits, 0 or 1)
    // bits  0-15 = errCode (reuses decodeAwaitErrCode)

    /**
     * Decode the timedOut flag from {@code cleat_await_promise}
     * (bits 16-23).
     *
     * @return true if the timeout expired before the promise resolved
     */
    public static boolean decodeAwaitPromiseTimedOut(long result) {
        return ((result >>> 16) & 0xFFL) != 0;
    }

    // ---- poll_signal result decoding ----
    // bits 32-63 = payloadLen
    // bits  8-15 = found flag (0x0100 = 256 if found)
    // bits  0-7  = errCode

    /**
     * Decode the payload length from {@code cleat_poll_signal}
     * (bits 32-63).
     */
    public static int decodePollSigPayloadLen(long result) {
        return (int) (result >>> 32);
    }

    /**
     * Decode the found flag from {@code cleat_poll_signal}
     * (bits 8-15).  Returns true if the value equals 0x0100 (256).
     */
    public static boolean decodePollSigFound(long result) {
        return ((result >>> 8) & 0xFFL) == 0x0100;
    }

    /**
     * Decode the error code from {@code cleat_poll_signal}
     * (bits 0-7).
     */
    public static int decodePollSigErrCode(long result) {
        return (int) (result & 0xFFL);
    }

    // ---- poll_cancellation result decoding ----
    // bits 32-63 = reasonLen
    // bits  0-31 = cancelled (non-zero = true)

    /**
     * Decode the cancelled flag from {@code cleat_poll_cancellation}
     * (bits 0-31).  Returns true if non-zero.
     */
    public static boolean decodePollCancelCancelled(long result) {
        return (result & 0xFFFFFFFFL) != 0;
    }

    /**
     * Decode the cancellation reason length from
     * {@code cleat_poll_cancellation} (bits 32-63).
     */
    public static int decodePollCancelReasonLen(long result) {
        return (int) (result >>> 32);
    }
}

File: crates/cleat-sdk/src/memory.rs
// Memory helpers for WASM linear memory.
// Matches the ABI from internal/host/memory.go.

/// Output buffer size constant (matches Go's outBufSize = 65536).
pub const OUT_BUF_SIZE: u32 = 65536;

/// Scratch region base offset in linear memory (10 MiB).
pub const SCRATCH_BASE: u32 = 10 * 1024 * 1024; // 0xA00000

/// Output buffer offset = scratch base + outBufSize.
pub const OUTPUT_OFFSET: u32 = SCRATCH_BASE + OUT_BUF_SIZE; // 0xA10000

/// Suspend sentinel (1 << 62). Matches runtime.go line 153.
pub const SUSPEND_SENTINEL: i64 = 1 << 62;

/// Bit 31 of a host-call result: the host is refusing this call because the
/// workflow is in a defer segment and the call would start new work.
///
/// Distinct from [`SUSPEND_SENTINEL`], which is bit 62 and is what the guest
/// returns to the host from an export. This one travels the other way -- host
/// to guest, inside an ordinary result word -- and bit 31 was chosen because it
/// is the one bit free in all six result layouts a call that can start fresh
/// work returns. See IMPROVEMENT-PLAN 3.84 and ABI.md.
///
/// The engine's copy is `callSuspendSentinel` in `engine/memory.go`. The two
/// must agree, and nothing in either language can see the other, so
/// `TestTheRustSDKAgreesOnTheStopBit` in `engine/` reads this file and pins the
/// value.
pub const SUSPEND_STOP_BIT: i64 = 1 << 31;

/// Read a host-written result out of a guest-owned buffer, bounded by the
/// buffer itself.
///
/// # Why this exists, and why it takes a slice
///
/// The host does not always write the buffer it reports a length for. On a
/// bad-parameter refusal `engine/imports.go` returns `errBadParam` =
/// `0xFFFFFFFF_00000001` from 54 sites, before the handler runs, and the guest
/// decodes a LENGTH out of the very bits that carry the sentinel:
///
/// | layout | decoded length | against a 65536-byte buffer |
/// |---|---|---|
/// | `decode_simple_result` | 4294967295 | overruns by ~65535x |
/// | `decode_cleat_call_result` | 16777215 | overruns by ~256x |
///
/// So a wrapper that reads its buffer on the error path -- which is the RIGHT
/// thing to do, because the host's real message is usually there -- reads far
/// out of bounds unless something bounds it. `read_string` cannot: it takes a
/// raw pointer and has no idea how big the region is.
///
/// This takes the `&[u8]` instead, so the capacity travels with the data and
/// there is no second argument to get wrong. It is also entirely safe code,
/// which is the point: the buffers in `host_calls.rs` are ordinary `Vec<u8>`,
/// and reading them back never needed `unsafe` at all.
///
/// The Java SDK's `readOutput` has clamped since it was written
/// (`Math.min(maxLen, OUT_BUF_SIZE)`), and Go's `hostErrMessage` bounds-checks
/// for exactly this reason -- see IMPROVEMENT-PLAN 3.200, which says so and
/// which the Rust side did not follow.
pub fn read_result(buf: &[u8], len: u32) -> String {
    let n = (len as usize).min(buf.len());
    if n == 0 {
        return String::new();
    }
    String::from_utf8_lossy(&buf[..n]).into_owned()
}

/// Read a string from WASM linear memory at (ptr, len).
/// Matches readWasmString in memory.go.
///
/// # Safety
///
/// `ptr` must point to a valid region of WASM linear memory at least `len` bytes long.
/// The caller must ensure that `ptr` is correctly aligned and that the memory region
/// is not concurrently mutated.
pub unsafe fn read_string(ptr: *const u8, len: u32) -> String {
    if len == 0 {
        return String::new();
    }
    let slice = unsafe { std::slice::from_raw_parts(ptr, len as usize) };
    String::from_utf8_lossy(slice).into_owned()
}

/// Write a string to WASM linear memory, truncating to max_len.
/// Returns the number of bytes written. Matches writeWasmString in memory.go.
///
/// # Safety
///
/// `ptr` must point to a valid region of WASM linear memory at least `max_len` bytes long.
/// The caller must ensure that `ptr` is correctly aligned and that the memory region
/// is not concurrently read or written during the copy.
pub unsafe fn write_string(ptr: *mut u8, max_len: u32, s: &str) -> u32 {
    let data = s.as_bytes();
    let len = (data.len() as u32).min(max_len);
    if len > 0 {
        unsafe {
            std::ptr::copy_nonoverlapping(data.as_ptr(), ptr, len as usize);
        }
    }
    len
}

/// Decode the export result: (errCode, actualLen).
/// Matches decodeExportResult in memory.go lines 68-70.
pub fn decode_export_result(result: u64) -> (u32, u32) {
    let err_code = (result & 0xFFFF_FFFF) as u32;
    let actual_len = (result >> 32) as u32;
    (err_code, actual_len)
}

/// Encode the export result for return from an export function.
/// Matches the convention: low 32 bits = errCode, high 32 bits = actualLen.
pub fn encode_export_result(err_code: u32, actual_len: u32) -> i64 {
    ((actual_len as u64) << 32 | (err_code as u64)) as i64
}

/// Pack a cleat_call result. Matches packDurableCallResult in memory.go lines 50-52.
/// bits 40-63 = responseLen (24 bits)
/// bits 8-39  = callErrorCode (32 bits)
/// bits 0-7   = errCode (8 bits)
pub fn decode_cleat_call_result(result: i64) -> (u32, u32, u8) {
    let r = result as u64;
    let response_len = ((r >> 40) & 0xFF_FFFF) as u32;
    let call_error_code = ((r >> 8) & 0xFFFF_FFFF) as u32;
    let err_code = (r & 0xFF) as u8;
    (response_len, call_error_code, err_code)
}

/// Pack a simple result: bits 32-63 = extra, bits 0-7 = errCode.
/// Matches packSimpleResult in memory.go lines 56-62.
pub fn decode_simple_result(result: i64) -> (u32, u8) {
    let r = result as u64;
    let extra = (r >> 32) as u32;
    let err_code = (r & 0xFF) as u8;
    (extra, err_code)
}

/// Decode sleep result. Matches engine.go lines 588-590.
/// bits 56-63 = status (0 = completed, 1 = suspend)
/// bits 0-55  = durationMs
pub const SLEEP_STATUS_COMPLETED: u8 = 0;
pub const SLEEP_STATUS_SUSPEND: u8 = 1;

pub fn decode_sleep_result(result: i64) -> (u8, i64) {
    let r = result as u64;
    let status = ((r >> 56) & 0xFF) as u8;
    // Mask to 56 bits for signed value
    let duration_ms = (r & 0x00FF_FFFF_FFFF_FFFF) as i64;
    (status, duration_ms)
}

/// Decode await_signals result. Matches engine.go lines 592-598.
/// bits 48-63 = sigNameLen (16 bits)
/// bits 32-47 = payloadLen (16 bits)
/// bits 16-23 = timedOut flag (1 byte)
/// bits 0-15  = errCode (16 bits)
pub fn decode_await_signals_result(result: i64) -> (u16, u16, bool, u16) {
    let r = result as u64;
    let sig_name_len = ((r >> 48) & 0xFFFF) as u16;
    let payload_len = ((r >> 32) & 0xFFFF) as u16;
    let timed_out = ((r >> 16) & 0xFFFF) != 0;
    let err_code = (r & 0xFFFF) as u16;
    (sig_name_len, payload_len, timed_out, err_code)
}

/// Decode await_promise result. ABI 2.21.
/// bits 32-63 = resultLen (32 bits)
/// bits 16-23 = timedOut flag (1 byte)
/// bits 0-15  = errCode (16 bits)
pub fn decode_await_promise_result(result: i64) -> (u32, bool, u16) {
    let r = result as u64;
    let result_len = ((r >> 32) & 0xFFFF_FFFF) as u32;
    let timed_out = ((r >> 16) & 0xFF) != 0;
    let err_code = (r & 0xFFFF) as u16;
    (result_len, timed_out, err_code)
}

/// PollSignal found flag (matches engine.go line 500: 0x0100).
pub const POLL_SIGNAL_FOUND: u32 = 0x0100;

/// Decode poll_signal result. Matches engine.go lines 500-504.
/// bits 32-63 = payloadLen
/// bits 8-15  = found flag (0x0100)
/// bits 0-7   = errCode
pub fn decode_poll_signal_result(result: i64) -> (u32, bool, u8) {
    let r = result as u64;
    let payload_len = (r >> 32) as u32;
    let flags = (r & 0xFFFF_FFFF) as u32;
    let err_code = (flags & 0xFF) as u8;
    let found = (flags >> 8) != 0;
    (payload_len, found, err_code)
}

/// Decode poll_cancellation result. Matches engine.go lines 488, 491.
/// bits 32-63 = reasonLen
/// bits 0-7   = cancelled flag (1 = cancelled)
pub fn decode_poll_cancellation_result(result: i64) -> (u32, bool) {
    let r = result as u64;
    let reason_len = (r >> 32) as u32;
    let cancelled = (r & 0xFFFF_FFFF) != 0;
    (reason_len, cancelled)
}

/// Decode get_scope result: upper 32 bits = objTypeLen, lower 32 bits = instKeyLen.
pub fn decode_get_scope_result(result: i64) -> (u32, u32) {
    let r = result as u64;
    let obj_type_len = (r >> 32) as u32;
    let inst_key_len = (r & 0xFFFF_FFFF) as u32;
    (obj_type_len, inst_key_len)
}

pub fn decode_incr_state_result(result: i64) -> (i64, u8) {
    let err_code = (result & 0xFF) as u8;
    let new_value = result >> 8;
    (new_value, err_code)
}

pub fn decode_has_state_result(result: i64) -> bool {
    result != 0
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The bad-parameter sentinel decodes to a length no buffer satisfies.
    ///
    /// This is the whole reason `read_result` exists, so it is asserted against
    /// the sentinel's real value rather than against a made-up large number.
    /// `engine/imports.go` returns `errBadParam` = 0xFFFFFFFF_00000001 from 54
    /// sites when it cannot read a guest string -- BEFORE the handler runs, so
    /// nothing has been written to the buffer at all -- and both result layouts
    /// decode a length out of the bits carrying the sentinel.
    #[test]
    fn err_bad_param_decodes_to_a_length_no_buffer_satisfies() {
        let bad = 0xFFFFFFFF_00000001u64 as i64;

        let (simple_len, err) = decode_simple_result(bad);
        assert_eq!(err, 1, "the low byte of errBadParam is 1");
        assert_eq!(simple_len, 4_294_967_295);
        assert!(
            simple_len as usize > OUT_BUF_SIZE as usize,
            "decode_simple_result gives {simple_len} against a {OUT_BUF_SIZE}-byte buffer"
        );

        let (call_len, _call_err, err2) = decode_cleat_call_result(bad);
        assert_eq!(err2, 1);
        assert_eq!(call_len, 16_777_215);
        assert!(
            call_len as usize > OUT_BUF_SIZE as usize,
            "decode_cleat_call_result gives {call_len} against a {OUT_BUF_SIZE}-byte buffer"
        );
    }

    /// read_result clamps to the buffer rather than to the reported length.
    ///
    /// Falsification: `String::from_utf8_lossy(&buf[..len as usize])` in place
    /// of the clamp panics here with a slice-index error -- in the SDK proper,
    /// where the read is from a raw pointer, the same mistake is an
    /// out-of-bounds READ rather than a panic, which is why this is checked on
    /// the safe helper.
    #[test]
    fn read_result_clamps_a_bogus_length_to_the_buffer() {
        let buf = vec![b'x'; 16];

        let (bogus, _) = decode_simple_result(0xFFFFFFFF_00000001u64 as i64);
        assert_eq!(read_result(&buf, bogus).len(), 16);

        // and the ordinary cases still behave
        assert_eq!(read_result(&buf, 4), "xxxx");
        assert_eq!(read_result(&buf, 0), "");
        assert_eq!(read_result(&[], 99), "");
    }

    /// `host_calls.rs` must not read a host-reported length from a raw pointer.
    ///
    /// `read_string` takes a `*const u8` and cannot bound anything, so every
    /// call site that passes a host-reported length is an out-of-bounds read
    /// waiting for a bad-parameter refusal. All 40 sites in `host_calls.rs`
    /// were converted to `read_result`; this stops the 41st being written.
    ///
    /// `read_string` itself is NOT deleted and must not be: `cleat-macro`'s
    /// generated entry point calls it on `(args_ptr, args_len)` handed in by
    /// the host, where there is no slice to bound against and the host is the
    /// one that chose the length. That is a different situation from reading
    /// back a buffer the guest allocated, and conflating them is what this
    /// test exists to prevent.
    ///
    /// Reads the file rather than grepping the repo, so it cannot be satisfied
    /// by a comment elsewhere and needs no tooling to run.
    #[test]
    fn host_calls_reads_every_buffer_through_read_result() {
        let src = include_str!("host_calls.rs");

        let unbounded = src
            .lines()
            .enumerate()
            .filter(|(_, l)| {
                let t = l.trim_start();
                !t.starts_with("//") && !t.starts_with("///") && t.contains("read_string(")
            })
            .map(|(i, l)| format!("  {}: {}", i + 1, l.trim()))
            .collect::<Vec<_>>();

        assert!(
            unbounded.is_empty(),
            "host_calls.rs reads a buffer through read_string, which takes a raw \
             pointer and bounds nothing:\n{}\n\n\
             Use memory::read_result(&buf, len) instead. On a bad-parameter refusal the \
             host returns errBadParam without writing the buffer, and the length decodes \
             to 4294967295 (simple layout) or 16777215 (call layout) against a 65536-byte \
             buffer. See IMPROVEMENT-PLAN 3.244.",
            unbounded.join("\n")
        );

        // A narrow backstop, and worth being honest about its scope: a rename
        // or a moved file is caught by the COMPILER long before this runs
        // (include_str! fails on a missing path, and the tests below call
        // read_result by name), which was checked rather than assumed -- an
        // attempted rename produced four E0425s and never reached this
        // assertion.
        //
        // What it does still catch is the case the first assertion cannot: a
        // host_calls.rs that reads no buffers at all, where "no unbounded
        // reads" is true and vacuous.
        assert!(
            src.contains("read_result("),
            "host_calls.rs contains no read_result( call at all, so the check above passed \
             vacuously rather than because the reads are bounded."
        );
    }

    /// A length shorter than the buffer must NOT be rounded up to it.
    ///
    /// Clamping is a maximum, not a replacement: the host reports how much it
    /// actually wrote into a buffer that is almost always larger, so a helper
    /// that returned the whole buffer would append thousands of NUL bytes to
    /// every successful result. That failure passes the clamp test above.
    #[test]
    fn read_result_does_not_round_a_short_length_up() {
        let mut buf = vec![0u8; 64];
        buf[..5].copy_from_slice(b"hello");
        assert_eq!(read_result(&buf, 5), "hello");
    }
}
